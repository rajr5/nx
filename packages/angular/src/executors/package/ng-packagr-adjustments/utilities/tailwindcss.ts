import { logger } from '@nrwl/devkit';
import { appRootPath } from '@nrwl/tao/src/utils/app-root';
import { existsSync } from 'fs';
import { join, relative } from 'path';
import * as postcssImport from 'postcss-import';

export function getTailwindPostCssPluginsIfPresent(
  basePath: string,
  includePaths?: string[],
  watch?: boolean
) {
  // Try to find TailwindCSS configuration file in the project or workspace root.
  const tailwindConfigFile = 'tailwind.config.js';
  let tailwindConfigPath: string | undefined;
  for (const path of [basePath, appRootPath]) {
    const fullPath = join(path, tailwindConfigFile);
    if (existsSync(fullPath)) {
      tailwindConfigPath = fullPath;
      break;
    }
  }

  // Only load Tailwind CSS plugin if configuration file was found.
  if (!tailwindConfigPath) {
    return [];
  }

  let tailwindPackagePath: string | undefined;
  try {
    tailwindPackagePath = require.resolve('tailwindcss');
  } catch {
    const relativeTailwindConfigPath = relative(
      appRootPath,
      tailwindConfigPath
    );
    logger.warn(
      `Tailwind CSS configuration file found (${relativeTailwindConfigPath})` +
        ` but the 'tailwindcss' package is not installed.` +
        ` To enable Tailwind CSS, please install the 'tailwindcss' package.`
    );

    return [];
  }

  if (!tailwindPackagePath) {
    return [];
  }

  if (process.env['TAILWIND_MODE'] === undefined) {
    process.env['TAILWIND_MODE'] = watch ? 'watch' : 'build';
  }

  return [
    postcssImport({
      addModulesDirectories: includePaths ?? [],
      resolve: (url: string) => (url.startsWith('~') ? url.substr(1) : url),
    }),
    require(tailwindPackagePath)({ config: tailwindConfigPath }),
  ];
}
