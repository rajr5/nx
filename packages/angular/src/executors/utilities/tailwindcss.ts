import { logger } from '@nrwl/devkit';
import { workspaceRoot } from '@nrwl/devkit';
import { existsSync } from 'fs';
import { join, relative } from 'path';
import * as postcssImport from 'postcss-import';

export interface TailwindSetup {
  tailwindConfigPath: string;
  tailwindPackagePath: string;
}

export const tailwindDirectives = [
  '@tailwind',
  '@apply',
  '@layer',
  '@variants',
  '@responsive',
  '@screen',
];

export function getTailwindSetup(
  basePath: string,
  tailwindConfig?: string
): TailwindSetup | undefined {
  let tailwindConfigPath = tailwindConfig;

  if (!tailwindConfigPath) {
    tailwindConfigPath = getTailwindConfigPath(basePath, workspaceRoot);
  }

  // Only load Tailwind CSS plugin if configuration file was found.
  if (!tailwindConfigPath) {
    return undefined;
  }

  let tailwindPackagePath: string | undefined;
  try {
    tailwindPackagePath = require.resolve('tailwindcss');
  } catch {
    const relativeTailwindConfigPath = relative(
      workspaceRoot,
      tailwindConfigPath
    );
    logger.warn(
      `Tailwind CSS configuration file found (${relativeTailwindConfigPath})` +
        ` but the 'tailwindcss' package is not installed.` +
        ` To enable Tailwind CSS, please install the 'tailwindcss' package.`
    );

    return undefined;
  }

  if (!tailwindPackagePath) {
    return undefined;
  }

  return { tailwindConfigPath, tailwindPackagePath };
}

function getTailwindConfigPath(
  projectRoot: string,
  workspaceRoot: string
): string | undefined {
  // valid tailwind config files https://github.com/tailwindlabs/tailwindcss/blob/master/src/util/resolveConfigPath.js#L46
  const tailwindConfigFiles = ['tailwind.config.js', 'tailwind.config.cjs'];

  for (const basePath of [projectRoot, workspaceRoot]) {
    for (const configFile of tailwindConfigFiles) {
      const fullPath = join(basePath, configFile);
      if (existsSync(fullPath)) {
        return fullPath;
      }
    }
  }

  return undefined;
}

export function getTailwindPostCssPlugins(
  { tailwindConfigPath, tailwindPackagePath }: TailwindSetup,
  includePaths?: string[],
  watch?: boolean
) {
  if (process.env['TAILWIND_MODE'] === undefined) {
    process.env['TAILWIND_MODE'] = watch ? 'watch' : 'build';
  }

  return [
    postcssImport({
      addModulesDirectories: includePaths ?? [],
      resolve: (url: string) => (url.startsWith('~') ? url.substring(1) : url),
    }),
    require(tailwindPackagePath)({ config: tailwindConfigPath }),
  ];
}
