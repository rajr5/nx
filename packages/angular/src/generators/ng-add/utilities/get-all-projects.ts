import { getProjects, Tree } from '@nrwl/devkit';
import { MigrationProjectConfiguration, WorkspaceProjects } from './types';

export function getAllProjects(tree: Tree): WorkspaceProjects {
  const projects = getProjects(tree);
  const result: WorkspaceProjects = {
    apps: [],
    libs: [],
  };

  for (const [name, project] of projects) {
    const migrationProject: MigrationProjectConfiguration = {
      config: project,
      name,
    };

    if (project.projectType === 'library') {
      result.libs.push(migrationProject);
    } else {
      result.apps.push(migrationProject);
    }
  }

  return result;
}
