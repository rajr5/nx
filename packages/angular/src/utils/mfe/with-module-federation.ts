import {
  getNpmPackageSharedConfig,
  SharedLibraryConfig,
  sharePackages,
  shareWorkspaceLibraries,
} from './mfe-webpack';
import {
  createProjectGraphAsync,
  ProjectGraph,
  readCachedProjectGraph,
  workspaceRoot,
  Workspaces,
} from '@nrwl/devkit';
import {
  getRootTsConfigPath,
  readTsConfig,
} from '@nrwl/workspace/src/utilities/typescript';
import { ParsedCommandLine } from 'typescript';
import { readWorkspaceJson } from 'nx/src/project-graph/file-utils';
import { readRootPackageJson } from './utils';
import { extname, join } from 'path';
import ModuleFederationPlugin = require('webpack/lib/container/ModuleFederationPlugin');

export type MFERemotes = string[] | [remoteName: string, remoteUrl: string][];

type SharedFunction = (
  libraryName: string,
  sharedConfig: SharedLibraryConfig
) => SharedLibraryConfig | false;
type AdditionalSharedConfig = Array<
  | string
  | [libraryName: string, sharedConfig: SharedLibraryConfig]
  | { libraryName: string; sharedConfig: SharedLibraryConfig }
>;

export interface MFEConfig {
  name: string;
  remotes?: MFERemotes;
  exposes?: Record<string, string>;
  shared?: SharedFunction;
  additionalShared?: AdditionalSharedConfig;
}

function collectDependencies(
  projectGraph: ProjectGraph,
  name: string,
  dependencies = {
    workspaceLibraries: new Set<string>(),
    npmPackages: new Set<string>(),
  },
  seen: Set<string> = new Set()
): {
  workspaceLibraries: Set<string>;
  npmPackages: Set<string>;
} {
  if (seen.has(name)) {
    return dependencies;
  }
  seen.add(name);

  (projectGraph.dependencies[name] ?? []).forEach((dependency) => {
    if (dependency.target.startsWith('npm:')) {
      dependencies.npmPackages.add(dependency.target.replace('npm:', ''));
    } else {
      dependencies.workspaceLibraries.add(dependency.target);
      collectDependencies(projectGraph, dependency.target, dependencies, seen);
    }
  });

  return dependencies;
}

function mapWorkspaceLibrariesToTsConfigImport(workspaceLibraries: string[]) {
  const { projects } = new Workspaces(
    workspaceRoot
  ).readWorkspaceConfiguration();

  const tsConfigPath = process.env.NX_TSCONFIG_PATH ?? getRootTsConfigPath();
  const tsConfig: ParsedCommandLine = readTsConfig(tsConfigPath);

  const tsconfigPathAliases: Record<string, string[]> = tsConfig.options?.paths;

  if (!tsconfigPathAliases) {
    return workspaceLibraries;
  }

  const mappedLibraries = [];
  for (const lib of workspaceLibraries) {
    const sourceRoot = projects[lib].sourceRoot;
    let found = false;

    for (const [key, value] of Object.entries(tsconfigPathAliases)) {
      if (value.find((p) => p.startsWith(sourceRoot))) {
        mappedLibraries.push(key);
        found = true;
        break;
      }
    }

    if (!found) {
      mappedLibraries.push(lib);
    }
  }

  return mappedLibraries;
}

async function getDependentPackagesForProject(
  projectGraph: ProjectGraph,
  name: string
) {
  const { npmPackages, workspaceLibraries } = collectDependencies(
    projectGraph,
    name
  );

  return {
    workspaceLibraries: mapWorkspaceLibrariesToTsConfigImport([
      ...workspaceLibraries,
    ]),
    npmPackages: [...npmPackages],
  };
}

function determineRemoteUrl(remote: string) {
  const workspace = readWorkspaceJson();
  let publicHost = '';
  try {
    publicHost = workspace.projects[remote].targets.serve.options.publicHost;
  } catch (error) {
    throw new Error(
      `Cannot automatically determine URL of remote (${remote}). Looked for property "publicHost" in the project's "serve" target.\n
      You can also use the tuple syntax in your webpack config to configure your remotes. e.g. \`remotes: [['remote1', 'http://localhost:4201']]\``
    );
  }
  return `${
    publicHost.endsWith('/') ? publicHost.slice(0, -1) : publicHost
  }/remoteEntry.mjs`;
}

function mapRemotes(remotes: MFERemotes) {
  const mappedRemotes = {};

  for (const remote of remotes) {
    if (Array.isArray(remote)) {
      const [remoteName, remoteLocation] = remote;
      const remoteLocationExt = extname(remoteLocation);
      mappedRemotes[remoteName] = ['.js', '.mjs'].includes(remoteLocationExt)
        ? remoteLocation
        : join(remoteLocation, 'remoteEntry.mjs');
    } else if (typeof remote === 'string') {
      mappedRemotes[remote] = determineRemoteUrl(remote);
    }
  }

  return mappedRemotes;
}

function applySharedFunction(
  sharedConfig: Record<string, SharedLibraryConfig>,
  sharedFn: SharedFunction | undefined
): void {
  if (!sharedFn) {
    return;
  }

  for (const [libraryName, library] of Object.entries(sharedConfig)) {
    const mappedDependency = sharedFn(libraryName, library);
    if (mappedDependency === false) {
      delete sharedConfig[libraryName];
      continue;
    } else if (!mappedDependency) {
      continue;
    }

    sharedConfig[libraryName] = mappedDependency;
  }
}

function addStringDependencyToSharedConfig(
  sharedConfig: Record<string, SharedLibraryConfig>,
  dependency: string,
  projectGraph: ProjectGraph
): void {
  if (projectGraph.nodes[dependency]) {
    sharedConfig[dependency] = { requiredVersion: false };
  } else if (projectGraph.externalNodes?.[dependency]) {
    const pkgJson = readRootPackageJson();
    const config = getNpmPackageSharedConfig(
      dependency,
      pkgJson.dependencies?.[dependency] ??
        pkgJson.devDependencies?.[dependency]
    );

    if (!config) {
      return;
    }

    sharedConfig[dependency] = config;
  } else {
    throw new Error(
      `The specified dependency "${dependency}" in the additionalShared configuration does not exist in the project graph. ` +
        `Please check your additionalShared configuration and make sure you are including valid workspace projects or npm packages.`
    );
  }
}

function applyAdditionalShared(
  sharedConfig: Record<string, SharedLibraryConfig>,
  additionalShared: AdditionalSharedConfig | undefined,
  projectGraph: ProjectGraph
): void {
  if (!additionalShared) {
    return;
  }

  for (const shared of additionalShared) {
    if (typeof shared === 'string') {
      addStringDependencyToSharedConfig(sharedConfig, shared, projectGraph);
    } else if (Array.isArray(shared)) {
      sharedConfig[shared[0]] = shared[1];
    } else if (typeof shared === 'object') {
      sharedConfig[shared.libraryName] = shared.sharedConfig;
    }
  }
}

function applyDefaultEagerPackages(
  sharedConfig: Record<string, SharedLibraryConfig>
) {
  const DEFAULT_PACKAGES_TO_LOAD_EAGERLY = ['@angular/localize/init'];
  for (const pkg of DEFAULT_PACKAGES_TO_LOAD_EAGERLY) {
    if (sharedConfig[pkg]) {
      return {
        ...sharedConfig,
        eager: true,
      };
    }
  }
}

export async function withModuleFederation(options: MFEConfig) {
  const DEFAULT_NPM_PACKAGES_TO_AVOID = ['zone.js', '@nrwl/angular/mfe'];

  let projectGraph: ProjectGraph<any>;
  try {
    projectGraph = readCachedProjectGraph();
  } catch (e) {
    projectGraph = await createProjectGraphAsync();
  }

  const dependencies = await getDependentPackagesForProject(
    projectGraph,
    options.name
  );
  const sharedLibraries = shareWorkspaceLibraries(
    dependencies.workspaceLibraries
  );

  const npmPackages = sharePackages(
    dependencies.npmPackages.filter(
      (pkg) => !DEFAULT_NPM_PACKAGES_TO_AVOID.includes(pkg)
    )
  );

  DEFAULT_NPM_PACKAGES_TO_AVOID.forEach((pkgName) => {
    if (pkgName in npmPackages) {
      delete npmPackages[pkgName];
    }
  });

  const sharedDependencies = {
    ...sharedLibraries.getLibraries(),
    ...npmPackages,
  };

  applyDefaultEagerPackages(sharedDependencies);
  applySharedFunction(sharedDependencies, options.shared);
  applyAdditionalShared(
    sharedDependencies,
    options.additionalShared,
    projectGraph
  );

  const mappedRemotes =
    !options.remotes || options.remotes.length === 0
      ? {}
      : mapRemotes(options.remotes);

  return (config) => ({
    ...(config ?? {}),
    output: {
      ...(config.output ?? {}),
      uniqueName: options.name,
      publicPath: 'auto',
    },
    optimization: {
      ...(config.optimization ?? {}),
      runtimeChunk: false,
    },
    resolve: {
      ...(config.resolve ?? {}),
      alias: {
        ...(config.resolve?.alias ?? {}),
        ...sharedLibraries.getAliases(),
      },
    },
    experiments: {
      ...(config.experiments ?? {}),
      outputModule: true,
    },
    plugins: [
      ...(config.plugins ?? []),
      new ModuleFederationPlugin({
        name: options.name,
        filename: 'remoteEntry.mjs',
        exposes: options.exposes,
        remotes: mappedRemotes,
        shared: {
          ...sharedDependencies,
        },
        library: {
          type: 'module',
        },
      }),
      sharedLibraries.getReplacementPlugin(),
    ],
  });
}
