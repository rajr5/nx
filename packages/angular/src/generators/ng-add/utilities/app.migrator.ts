import {
  joinPathFragments,
  offsetFromRoot,
  readJson,
  TargetConfiguration,
  Tree,
  updateJson,
  updateProjectConfiguration,
} from '@nrwl/devkit';
import { hasRulesRequiringTypeChecking } from '@nrwl/linter';
import { convertToNxProjectGenerator } from '@nrwl/workspace/generators';
import { getRootTsConfigPathInTree } from '@nrwl/workspace/src/utilities/typescript';
import { basename } from 'path';
import { GeneratorOptions } from '../schema';
import { E2eProjectMigrator } from './e2e.migrator';
import { ProjectMigrator } from './project.migrator';
import { MigrationProjectConfiguration, ValidationResult } from './types';

export class AppMigrator extends ProjectMigrator {
  private e2eMigrator: E2eProjectMigrator;

  constructor(
    tree: Tree,
    options: GeneratorOptions,
    project: MigrationProjectConfiguration
  ) {
    super(tree, options, project, 'apps');

    this.e2eMigrator = new E2eProjectMigrator(tree, options, project);
  }

  async migrate(): Promise<void> {
    await this.e2eMigrator.migrate();

    this.moveProjectFiles();
    await this.updateProjectConfiguration();
    this.updateTsConfigs();
    this.updateEsLintConfig();
  }

  validate(): ValidationResult {
    const result: ValidationResult = [];

    if (
      this.projectConfig.targets.lint &&
      this.projectConfig.targets.lint.executor !==
        '@angular-eslint/builder:lint'
    ) {
      result.push({
        message: `The "lint" target is using an unsupported builder "${this.projectConfig.targets.lint.executor}".`,
        hint: `The supported builder is "@angular-eslint/builder:lint".`,
      });
    }

    result.push(...(this.e2eMigrator.validate() ?? []));

    return result.length ? result : null;
  }

  private moveProjectFiles(): void {
    // it is not required to have a browserslist
    this.moveProjectRootFile(
      joinPathFragments(this.project.oldRoot, 'browserslist'),
      false
    );
    this.moveProjectRootFile(
      joinPathFragments(this.project.oldRoot, '.browserslistrc'),
      false
    );

    this.moveFilePathsFromTargetToProjectRoot(
      this.projectConfig.targets.build,
      ['tsConfig', 'webWorkerTsConfig', 'ngswConfigPath']
    );

    if (this.projectConfig.targets.test) {
      this.moveFilePathsFromTargetToProjectRoot(
        this.projectConfig.targets.test,
        ['karmaConfig', 'tsConfig', 'webWorkerTsConfig']
      );
    } else {
      // there could still be a karma.conf.js file in the root
      // so move to new location
      const karmaConfig = joinPathFragments(
        this.project.oldRoot,
        'karma.conf.js'
      );
      if (this.tree.exists(karmaConfig)) {
        this.logger.info(
          'No "test" target was found, but a root Karma config file was found in the project root. The file will be moved to the new location.'
        );
        this.moveProjectRootFile(karmaConfig);
      }
    }

    if (this.projectConfig.targets.server) {
      this.moveProjectRootFile(
        this.projectConfig.targets.server.options.tsConfig
      );
    }

    if (this.projectConfig.targets.lint) {
      this.moveProjectRootFile(
        this.projectConfig.targets.lint.options.eslintConfig ??
          joinPathFragments(this.project.oldRoot, '.eslintrc.json')
      );
    } else {
      // there could still be a .eslintrc.json file in the root
      // so move to new location
      const eslintConfig = joinPathFragments(
        this.project.oldRoot,
        '.eslintrc.json'
      );
      if (this.tree.exists(eslintConfig)) {
        this.logger.info(
          'No "lint" target was found, but an ESLint config file was found in the project root. The file will be moved to the new location.'
        );
        this.moveProjectRootFile(eslintConfig);
      }
    }

    this.moveDir(this.project.oldSourceRoot, this.project.newSourceRoot);
  }

  private async updateProjectConfiguration(): Promise<void> {
    this.projectConfig.root = this.project.newRoot;
    this.projectConfig.sourceRoot = this.project.newSourceRoot;

    this.convertBuildOptions(this.projectConfig.targets.build.options);
    Object.values(
      this.projectConfig.targets.build.configurations ?? {}
    ).forEach((config) => this.convertBuildOptions(config));

    if (this.projectConfig.targets.test) {
      const testOptions = this.projectConfig.targets.test.options;
      testOptions.main =
        testOptions.main && this.convertAsset(testOptions.main);
      testOptions.polyfills =
        testOptions.polyfills && this.convertAsset(testOptions.polyfills);
      testOptions.tsConfig = joinPathFragments(
        this.project.newRoot,
        'tsconfig.spec.json'
      );
      testOptions.karmaConfig = joinPathFragments(
        this.project.newRoot,
        'karma.conf.js'
      );
      testOptions.assets =
        testOptions.assets &&
        testOptions.assets.map((asset) => this.convertAsset(asset));
      testOptions.styles =
        testOptions.styles &&
        testOptions.styles.map((style) => this.convertAsset(style));
      testOptions.scripts =
        testOptions.scripts &&
        testOptions.scripts.map((script) => this.convertAsset(script));
    }

    if (this.projectConfig.targets.lint) {
      this.projectConfig.targets.lint.executor = '@nrwl/linter:eslint';
      const lintOptions = this.projectConfig.targets.lint.options;
      lintOptions.eslintConfig =
        lintOptions.eslintConfig &&
        joinPathFragments(
          this.project.newRoot,
          basename(lintOptions.eslintConfig)
        );
      lintOptions.lintFilePatterns =
        lintOptions.lintFilePatterns &&
        lintOptions.lintFilePatterns.map((pattern) =>
          this.convertPath(pattern)
        );

      const eslintConfigPath =
        lintOptions.eslintConfig ??
        joinPathFragments(this.project.newRoot, '.eslintrc.json');
      const eslintConfig = readJson(this.tree, eslintConfigPath);
      if (hasRulesRequiringTypeChecking(eslintConfig)) {
        lintOptions.hasTypeAwareRules = true;
      }
    }

    if (this.projectConfig.targets.server) {
      const serverOptions = this.projectConfig.targets.server.options;
      this.convertServerOptions(serverOptions);
      Object.values(this.projectConfig.targets.server.configurations).forEach(
        (config) => this.convertServerOptions(config)
      );
    }

    updateProjectConfiguration(this.tree, this.project.name, {
      ...this.projectConfig,
    });

    await convertToNxProjectGenerator(this.tree, {
      project: this.project.name,
      skipFormat: true,
    });
  }

  private updateTsConfigs(): void {
    const rootTsConfigFile = getRootTsConfigPathInTree(this.tree);
    const projectOffsetFromRoot = offsetFromRoot(this.projectConfig.root);

    this.updateTsConfigFile(
      this.projectConfig.targets.build.options.tsConfig,
      rootTsConfigFile,
      projectOffsetFromRoot
    );

    if (this.projectConfig.targets.test) {
      this.updateTsConfigFile(
        this.projectConfig.targets.test.options.tsConfig,
        rootTsConfigFile,
        projectOffsetFromRoot
      );
    }

    if (this.projectConfig.targets.server) {
      this.updateTsConfigFile(
        this.projectConfig.targets.server.options.tsConfig,
        rootTsConfigFile,
        projectOffsetFromRoot
      );
    }
  }

  private updateEsLintConfig(): void {
    const eslintConfigPath =
      this.projectConfig.targets.lint?.options?.eslintConfig ??
      joinPathFragments(this.project.newRoot, '.eslintrc.json');

    if (!this.tree.exists(eslintConfigPath)) {
      return;
    }

    updateJson(this.tree, eslintConfigPath, (json) => {
      delete json.root;
      json.ignorePatterns = ['!**/*'];

      const rootEsLintConfigRelativePath = joinPathFragments(
        offsetFromRoot(this.projectConfig.root),
        '.eslintrc.json'
      );
      if (Array.isArray(json.extends)) {
        json.extends = json.extends.map((extend: string) =>
          this.convertEsLintConfigExtendToNewPath(eslintConfigPath, extend)
        );

        // it might have not been extending from the root config, make sure it does
        if (!json.extends.includes(rootEsLintConfigRelativePath)) {
          json.extends.push(rootEsLintConfigRelativePath);
        }
      } else {
        json.extends = rootEsLintConfigRelativePath;
      }

      json.overrides?.forEach((override) => {
        if (!override.parserOptions?.project) {
          return;
        }

        override.parserOptions.project = [
          `${this.projectConfig.root}/tsconfig.*?.json`,
        ];
      });

      return json;
    });
  }

  private convertBuildOptions(buildOptions: any): void {
    buildOptions.outputPath =
      buildOptions.outputPath &&
      joinPathFragments('dist', this.project.newRoot);
    buildOptions.index =
      buildOptions.index && this.convertAsset(buildOptions.index);
    buildOptions.main =
      buildOptions.main && this.convertAsset(buildOptions.main);
    buildOptions.polyfills =
      buildOptions.polyfills && this.convertAsset(buildOptions.polyfills);
    buildOptions.tsConfig =
      buildOptions.tsConfig &&
      joinPathFragments(this.project.newRoot, 'tsconfig.app.json');
    buildOptions.assets =
      buildOptions.assets &&
      buildOptions.assets.map((asset) => this.convertAsset(asset));
    buildOptions.styles =
      buildOptions.styles &&
      buildOptions.styles.map((style) => this.convertAsset(style));
    buildOptions.scripts =
      buildOptions.scripts &&
      buildOptions.scripts.map((script) => this.convertAsset(script));
    buildOptions.fileReplacements =
      buildOptions.fileReplacements &&
      buildOptions.fileReplacements.map((replacement: any) => ({
        replace: this.convertAsset(replacement.replace),
        with: this.convertAsset(replacement.with),
      }));
  }

  private convertServerOptions(serverOptions: any): void {
    serverOptions.outputPath =
      serverOptions.outputPath &&
      joinPathFragments('dist', 'apps', `${this.project.name}-server`);
    serverOptions.main =
      serverOptions.main && this.convertAsset(serverOptions.main);
    serverOptions.tsConfig =
      serverOptions.tsConfig &&
      joinPathFragments(this.project.newRoot, 'tsconfig.server.json');
    serverOptions.fileReplacements =
      serverOptions.fileReplacements &&
      serverOptions.fileReplacements.map((replacement) => ({
        replace: this.convertAsset(replacement.replace),
        with: this.convertAsset(replacement.with),
      }));
  }

  private moveFilePathsFromTargetToProjectRoot(
    target: TargetConfiguration,
    options: string[]
  ) {
    options.forEach((option) => {
      this.getTargetValuesForOption(target, option).forEach((path) => {
        this.moveProjectRootFile(path);
      });
    });
  }
}
