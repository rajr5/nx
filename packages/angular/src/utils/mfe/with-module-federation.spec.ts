jest.mock('fs');
jest.mock('@nrwl/devkit');
jest.mock('@nrwl/workspace/src/utilities/typescript');
jest.mock('nx/src/core/file-utils');
jest.mock('nx/src/shared/workspace');
import * as graph from '@nrwl/devkit';
import * as typescriptUtils from '@nrwl/workspace/src/utilities/typescript';
import * as workspace from 'nx/src/core/file-utils';
import * as fs from 'fs';

import { withModuleFederation } from './with-module-federation';

xdescribe('withModuleFederation', () => {
  afterEach(() => jest.clearAllMocks());
  it('should create a host config correctly', async () => {
    // ARRANGE
    (graph.readCachedProjectGraph as jest.Mock).mockReturnValue({
      dependencies: {
        host: [
          { target: 'npm:@angular/core' },
          { target: 'npm:zone.js' },
          { target: 'shared' },
        ],
      },
    });

    (workspace.readWorkspaceJson as jest.Mock).mockReturnValue({
      projects: {
        remote1: {
          targets: {
            serve: {
              options: {
                publicHost: 'http://localhost:4201',
              },
            },
          },
        },
      },
    });

    (fs.existsSync as jest.Mock).mockReturnValue(true);
    (fs.readFileSync as jest.Mock).mockReturnValue(
      JSON.stringify({
        dependencies: {
          '@angular/core': '~13.2.0',
        },
      })
    );

    (typescriptUtils.readTsConfig as jest.Mock).mockReturnValue({
      options: {
        paths: {
          shared: ['/libs/shared/src/index.ts'],
        },
      },
    });

    (graph.Workspaces as jest.Mock).mockReturnValue({
      readWorkspaceConfiguration: () => ({
        projects: {
          shared: {
            sourceRoot: '/libs/shared/src/',
          },
        },
      }),
    });

    // ACT
    const config = (
      await withModuleFederation({
        name: 'host',
        remotes: ['remote1'],
      })
    )({});

    // ASSERT
    expect(config.plugins).toMatchSnapshot();
  });

  it('should create a remote config correctly', async () => {
    // ARRANGE
    (graph.readCachedProjectGraph as jest.Mock).mockReturnValue({
      dependencies: {
        remote1: [
          { target: 'npm:@angular/core' },
          { target: 'npm:zone.js' },
          { target: 'shared' },
        ],
      },
    });

    (fs.existsSync as jest.Mock).mockReturnValue(true);
    (fs.readFileSync as jest.Mock).mockReturnValue(
      JSON.stringify({
        dependencies: {
          '@angular/core': '~13.2.0',
        },
      })
    );

    (typescriptUtils.readTsConfig as jest.Mock).mockReturnValue({
      options: {
        paths: {
          shared: ['/libs/shared/src/index.ts'],
        },
      },
    });

    (graph.Workspaces as jest.Mock).mockReturnValue({
      readWorkspaceConfiguration: () => ({
        projects: {
          shared: {
            sourceRoot: '/libs/shared/src/',
          },
        },
      }),
    });

    // ACT
    const config = (
      await withModuleFederation({
        name: 'remote1',
        exposes: { './Module': 'apps/remote1/src/module.ts' },
      })
    )({});

    // ASSERT
    expect(config.plugins).toMatchSnapshot();
  });

  it('should collect dependencies correctly', async () => {
    // ARRANGE
    (graph.readCachedProjectGraph as jest.Mock).mockReturnValue({
      dependencies: {
        remote1: [
          { target: 'npm:@angular/core' },
          { target: 'npm:zone.js' },
          { target: 'core' },
        ],
        core: [{ target: 'shared' }],
      },
    });

    (fs.existsSync as jest.Mock).mockReturnValue(true);
    (fs.readFileSync as jest.Mock).mockReturnValue(
      JSON.stringify({
        dependencies: {
          '@angular/core': '~13.2.0',
        },
      })
    );

    (typescriptUtils.readTsConfig as jest.Mock).mockReturnValue({
      options: {
        paths: {
          shared: ['/libs/shared/src/index.ts'],
          core: ['/libs/core/src/index.ts'],
        },
      },
    });

    (graph.Workspaces as jest.Mock).mockReturnValue({
      readWorkspaceConfiguration: () => ({
        projects: {
          shared: {
            sourceRoot: '/libs/shared/src/',
          },
          core: {
            sourceRoot: '/libs/core/src/',
          },
        },
      }),
    });

    // ACT
    const config = (
      await withModuleFederation({
        name: 'remote1',
        exposes: { './Module': 'apps/remote1/src/module.ts' },
      })
    )({});

    // ASSERT
    expect(config.plugins).toMatchSnapshot();
  });

  it('should map dependencies from project name to import name', async () => {
    // ARRANGE
    (graph.readCachedProjectGraph as jest.Mock).mockReturnValue({
      dependencies: {
        remote1: [
          { target: 'npm:@angular/core' },
          { target: 'npm:zone.js' },
          { target: 'core' },
        ],
        core: [{ target: 'shared' }],
      },
    });

    (fs.existsSync as jest.Mock).mockReturnValue(true);
    (fs.readFileSync as jest.Mock).mockReturnValue(
      JSON.stringify({
        dependencies: {
          '@angular/core': '~13.2.0',
        },
      })
    );

    (typescriptUtils.readTsConfig as jest.Mock).mockImplementation(() => ({
      options: {
        paths: {
          shared: ['/libs/shared/src/index.ts'],
          '@myorg/core': ['/libs/core/src/index.ts'],
        },
      },
    }));

    (graph.Workspaces as jest.Mock).mockReturnValue({
      readWorkspaceConfiguration: () => ({
        projects: {
          shared: {
            sourceRoot: '/libs/shared/src/',
          },
          core: {
            sourceRoot: '/libs/core/src/',
          },
        },
      }),
    });

    // ACT
    const config = (
      await withModuleFederation({
        name: 'remote1',
        exposes: { './Module': 'apps/remote1/src/module.ts' },
      })
    )({});

    // ASSERT
    expect(config.plugins).toMatchSnapshot();
  });
});
