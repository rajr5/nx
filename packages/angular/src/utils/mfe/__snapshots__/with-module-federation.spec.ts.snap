// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withModuleFederation should collect dependencies correctly 1`] = `
Array [
  ModuleFederationPlugin {
    "_options": Object {
      "exposes": Object {
        "./Module": "apps/remote1/src/module.ts",
      },
      "filename": "remoteEntry.mjs",
      "library": Object {
        "type": "module",
      },
      "name": "remote1",
      "remotes": Object {},
      "shared": Object {
        "@angular/core": Object {
          "requiredVersion": "~13.2.0",
          "singleton": true,
          "strictVersion": true,
        },
        "core": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
        "shared": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
      },
    },
  },
  NormalModuleReplacementPlugin {
    "newResource": [Function],
    "resourceRegExp": /\\./,
  },
]
`;

exports[`withModuleFederation should create a host config correctly 1`] = `
Array [
  ModuleFederationPlugin {
    "_options": Object {
      "exposes": undefined,
      "filename": "remoteEntry.mjs",
      "library": Object {
        "type": "module",
      },
      "name": "host",
      "remotes": Object {
        "remote1": "http:/localhost:4201/remoteEntry.mjs",
      },
      "shared": Object {
        "@angular/core": Object {
          "requiredVersion": "~13.2.0",
          "singleton": true,
          "strictVersion": true,
        },
        "shared": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
      },
    },
  },
  NormalModuleReplacementPlugin {
    "newResource": [Function],
    "resourceRegExp": /\\./,
  },
]
`;

exports[`withModuleFederation should create a remote config correctly 1`] = `
Array [
  ModuleFederationPlugin {
    "_options": Object {
      "exposes": Object {
        "./Module": "apps/remote1/src/module.ts",
      },
      "filename": "remoteEntry.mjs",
      "library": Object {
        "type": "module",
      },
      "name": "remote1",
      "remotes": Object {},
      "shared": Object {
        "@angular/core": Object {
          "requiredVersion": "~13.2.0",
          "singleton": true,
          "strictVersion": true,
        },
        "shared": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
      },
    },
  },
  NormalModuleReplacementPlugin {
    "newResource": [Function],
    "resourceRegExp": /\\./,
  },
]
`;

exports[`withModuleFederation should map dependencies from project name to import name 1`] = `
Array [
  ModuleFederationPlugin {
    "_options": Object {
      "exposes": Object {
        "./Module": "apps/remote1/src/module.ts",
      },
      "filename": "remoteEntry.mjs",
      "library": Object {
        "type": "module",
      },
      "name": "remote1",
      "remotes": Object {},
      "shared": Object {
        "@angular/core": Object {
          "requiredVersion": "~13.2.0",
          "singleton": true,
          "strictVersion": true,
        },
        "@myorg/core": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
        "shared": Object {
          "eager": undefined,
          "requiredVersion": false,
        },
      },
    },
  },
  NormalModuleReplacementPlugin {
    "newResource": [Function],
    "resourceRegExp": /\\./,
  },
]
`;
